### PR TITLE
ci: Add automated docker builds for opnode and l2 output submitter

### PR DESCRIPTION
**Pull request by mslipper**

### CODE REVIEW

File: ops/l2os-entrypoint.sh [hunk 1]
```diff
@@ -1,14 +0,0 @@
-#!/bin/sh
```
Comment (protolambda) [inline ops/l2os-entrypoint.sh:1]: I think it's nice to reduce code, and don't like the curl at startup, but why exactly does this PR remove these sh files?

File: ops/Dockerfile.l2os [hunk 2]
```diff
@@ -7,17 +7,13 @@ COPY go.mod /app/go.mod
...
 
 FROM alpine:3.15
 
-RUN apk add --no-cache curl jq
-
 COPY --from=builder /app/bin/l2os /usr/local/bin
-COPY --from=builder /app/l2os-entrypoint.sh /l2os-entrypoint.sh
 
-ENTRYPOINT ["/bin/sh", "/l2os-entrypoint.sh"]
+CMD ["l2os"]
```
Comment (protolambda) [inline ops/Dockerfile.l2os:19]: nice clean up

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      - checkout
+      - setup_remote_docker:
+          version: 19.03.13
+      - run:
+          name: Build
+          command: |
+            echo "$DOCKER_PASS" | docker login -u "$DOCKER_USERNAME" --password-stdin
+            docker build -t "ethereumoptimism/<<parameters.image-name>>:<<parameters.image-tag>>" -f <<parameters.dockerfile>> <<#parameters.target>>--target <<parameters.target>><</parameters.target>> .
+            docker push "ethereumoptimism/<<parameters.image-name>>:<<parameters.image-tag>>"
+
```
Comment (protolambda) [inline .circleci/config.yml:32]: Not sure how to test this part without access, seems to be the complicated part of the PR.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
+version: 2.1
+
+
+jobs:
+  build-dockerfile:
+    docker:
+      - image: cimg/base:2021.04
```
Comment (protolambda) [inline .circleci/config.yml:7]: should we try a newer image?

**Review by trianglesphere — COMMENTED** [review]
A couple small Qs. Generally looks fine.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+          context:
+            - optimism
+          filters:
+            branches:
+              only:
+                - main
+          image-name: l2os
+          image-tag: develop
+          dockerfile: ops/Dockerfile.l2os
+          name: Build L2 output submitter
```
Comment (maurelian) [inline .circleci/config.yml:57]: Friendly nudge to make use of the editorConfig file we have in this repo and the monorepo. I think you're on [IntelliJ](https://www.jetbrains.com/help/idea/editorconfig.html)?

File: ops/l2os-entrypoint.sh (hunk 1 above)
Comment (mslipper) [reply ops/l2os-entrypoint.sh:1]: The best practice for Docker containers is to treat them like executables. They should usually expose a single `CMD` that people can then pass env vars or CLI flags.

We violated this best practice in order to make sure that dependent services were online before starting the main executable in the container. The devnet-up script does this for us now, so we can remove these entrypoint scripts.

File: .circleci/config.yml (hunk 4 above)
Comment (mslipper) [reply .circleci/config.yml:7]: Yeah I can upgrade this.

[the remark this answers is not in the document]
Comment (mslipper) [reply .circleci/config.yml:7]: 👍 will upgrade.

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        description: Dockerfile to use
+        type: string
+    steps:
+      - checkout
+      - setup_remote_docker:
+          version: 19.03.13
+      - run:
+          name: Build
+          command: |
+            echo "$DOCKER_PASS" | docker login -u "$DOCKER_USERNAME" --password-stdin
```
[the remark this answers is not in the document]
Comment (mslipper) [reply .circleci/config.yml:29]: The CircleCI side injects these environment variables at runtime.

They can be leaked if we run untrusted code within CircleCI, which is why we have to whitelist untrusted PRs before running them.

File: .circleci/config.yml (hunk 3 above)
Comment (mslipper) [reply .circleci/config.yml:32]: Once this is merged I'll run the job to test this out. We've used this exact code elsewhere to push images, the only issue will be setting up the Docker context which happens in the CircleCI Console.

**Review by maurelian — APPROVED** [review]

**Pull request merged**